File: PSDatePicker-ObjC/PSDatePicker/PSDatePickerLib/PSDatePicker.m
//
//  PSDatePicker.m
//  PSDatePicker
//
//  Created by Pawan Kumar Singh on 26/01/15.
//  Copyright (c) 2015 Pawan Kumar Singh. All rights reserved.
//

#import "PSDatePicker.h"
@interface PSDatePicker() <UIPickerViewDataSource, UIPickerViewDelegate>

@property (nonatomic, assign) NSInteger day;
@property (nonatomic, assign) NSInteger month;
@property (nonatomic, assign) NSInteger year;
@property (nonatomic, strong) NSArray *monthArray;
@end

@implementation PSDatePicker

- (PSDatePicker *)initWithFrame: (CGRect)frame
{
    self = [super initWithFrame:frame];
    if(self)
    {
        self.monthArray = @[@"January", @"February", @"March", @"April", @"May", @"June", @"July", @"August", @"September", @"October", @"November", @"December"];
        self.datePickerMode = PSDatePickerModeDateAndTime;
    }
    return self;
}

- (PSDatePicker *)initWithPickerMode : (PSDatePickerMode)pickerMode
{
    switch (pickerMode) {
        case PSDatePickerModeMonthAndYear:
        {
            UIPickerView *picker = [[UIPickerView alloc] init];
            picker.tag = 1002;
            CGRect frame = picker.frame;
            self = [self initWithFrame:frame];
            [self initializeDateComponentsForDate:[NSDate date]];
            _day = 0;
            self.datePickerMode = PSDatePickerModeMonthAndYear;
            picker.dataSource = self;
            picker.delegate = self;
            [picker selectRow:[self currentMonthIndex] inComponent:0 animated:true];
            [picker selectRow:[self currentYearIndex] inComponent:1 animated:true];
            [self addSubview:picker];
        }
            break;
        case PSDatePickerModeDateAndMonth:
        {
            UIPickerView *picker = [[UIPickerView alloc] init];
            picker.tag = 1002;
            CGRect frame = picker.frame;
            self = [self initWithFrame: frame];
            [self initializeDateComponentsForDate:[NSDate date]];
            self.datePickerMode = PSDatePickerModeDateAndMonth;
            _year = 0;
            picker.dataSource = self;
            picker.delegate = self;
            [picker selectRow:[self todayDateIndex] inComponent:0 animated:true];
            [picker selectRow:[self currentMonthIndex] inComponent:1 animated:true];
            [self addSubview:picker];
        }
            break;
        default:
        {
            UIDatePicker *picker = [[UIDatePicker alloc] init];
            CGRect frame = picker.frame;
            self = [self initWithFrame:frame];
            picker.tag = 1002;
            picker.datePickerMode = [self getUIDatePickerMode:pickerMode];
            [self addSubview:picker];
        }
    }
    return self;
}

- (PSDatePicker *)init
{
    self = [self initWithPickerMode:PSDatePickerModeDateAndTime];
    return self;
}

-(NSString *)description
{
    NSMutableString *dateString = [[NSMutableString alloc] initWithString:@""];
    if (self.datePickerMode == PSDatePickerModeDateAndMonth || self.datePickerMode == PSDatePickerModeMonthAndYear)
    {
        if (_year != 0) {
            [dateString appendFormat:@"%04ld", _year];
        }
        if (_month != 0) {
            (dateString.length == 0) ? [dateString appendFormat:@"%02ld", _month] : [dateString appendFormat:@"-%02ld", _month];
        }
        if (_day != 0) {
            (dateString.length == 0) ? [dateString appendFormat:@"%02ld", _day] : [dateString appendFormat:@"-%02ld", _day];
        }
    }else {
        UIDatePicker *datePicker = (UIDatePicker *)[self viewWithTag:1002];
        [dateString appendString:datePicker.date.description];
    }
    return dateString;
}

- (void)initializeDateComponentsForDate: (NSDate *)date
{
    //getting date components for today's date
    NSCalendar *gregorianCalender  = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSCalendarUnit flags = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay;
    NSDateComponents *dateComponents = [gregorianCalender components:flags fromDate:[NSDate date]];
    
    _day = dateComponents.day;
    _month = dateComponents.month;
    _year = dateComponents.year;
    
}

#pragma mark - UIPickerViewDataSource -

// returns the number of 'columns' to display.
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 2;
}

// returns the number of rows in each component..
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    NSInteger number = 0;
    if (self.datePickerMode == PSDatePickerModeDateAndMonth)
    {
        if(component == 0)
        {
            number = [@[@31,@29,@31,@30,@31,@30,@31,@31,@30,@31,@30,@31][_month] integerValue];
        }else
        {
            number = 12;     // 12 months
        }
    }else if (self.datePickerMode == PSDatePickerModeMonthAndYear)
    {
        if(component == 0)
        {
            number = 12;     //12 months
        }else
        {
            number = 10000;  //10000 year. Same as default UIDatePicker if maximum and minimum date is not set
        }
    }
    return number;
}

//
- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    if (self.datePickerMode == PSDatePickerModeDateAndMonth)
    {
        if(component == 0)
        {
            //Date component is at 0
            _day = row+1;
            
        }else{
            //Month component is at 1
            _month = row+1;
            [pickerView reloadAllComponents];
        }
        
    }else if (self.datePickerMode == PSDatePickerModeMonthAndYear){
        if(component == 0)
        {
            //Month component is at 0
            _month = row+1;
        }else{
            //Year component is at 1
            _year = row+1;
        }
    }
}

//MARK: - UIPickerViewDelegate
- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    switch (self.datePickerMode)
    {
        case PSDatePickerModeDateAndMonth:
            return (component == 0) ? [NSString stringWithFormat:@"%2ld",row+1] : self.monthArray[row];
        case PSDatePickerModeMonthAndYear:
            return (component == 0) ? self.monthArray[row] : [NSString stringWithFormat:@"%ld",row+1];
        default:
            return @"";
    }
}

#pragma mark - Helper methods -
- (NSInteger)todayDateIndex
{
    NSCalendar *gregorianCalender  = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSCalendarUnit flags = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay;
    NSDateComponents *dateComponents = [gregorianCalender components:flags fromDate:[NSDate date]];
    return dateComponents.day-1;
}

- (NSInteger)currentMonthIndex
{
    NSCalendar *gregorianCalender  = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSCalendarUnit flags = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay;
    NSDateComponents *dateComponents = [gregorianCalender components:flags fromDate:[NSDate date]];
    return dateComponents.month-1;
}

- (NSInteger)currentYearIndex
{
    NSCalendar *gregorianCalender  = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian];
    NSCalendarUnit flags = NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitDay;
    NSDateComponents *dateComponents = [gregorianCalender components:flags fromDate:[NSDate date]];
    return dateComponents.year-1;
}

- (UIDatePickerMode)getUIDatePickerMode:(PSDatePickerMode)datePickerMode
{
    switch (datePickerMode) {
        case PSDatePickerModeTime:
            return UIDatePickerModeTime;
        case PSDatePickerModeDate:
            return UIDatePickerModeDate;
        case PSDatePickerModeDateAndTime:
            return UIDatePickerModeDateAndTime;
        case PSDatePickerModeCountDownTimer:
            return UIDatePickerModeCountDownTimer;
        default:
            return UIDatePickerModeDate;
    }
}

// if animated is YES, animate the wheels of time to display the new date
- (void)setDate:(NSDate *)date animated:(BOOL)animate
{
    
}

@end

File: PSDatePicker-Swift/PSDatePicker/PSDatePickerLib/PSDatePicker.swift
//
//  PSDatePicker.swift
//  SwiftPicker
//
//  Created by Pawan Kumar Singh on 01/11/14.
//  Copyright (c) 2014 Pawan Kumar Singh. All rights reserved.
//

import UIKit

class PSDatePicker : UIView, UIPickerViewDataSource, UIPickerViewDelegate {

    enum PSDatePickerMode : Int {
        
        case Time // Displays hour, minute, and optionally AM/PM designation depending on the locale setting (e.g. 6 | 53 | PM)
        case Date // Displays month, day, and year depending on the locale setting (e.g. November | 15 | 2007)
        case DateAndTime // Displays date, hour, minute, and optionally AM/PM designation depending on the locale setting (e.g. Wed Nov 15 | 6 | 53 | PM)
        case CountDownTimer // Displays hour and minute (e.g. 1 | 53)
        case MonthAndYear   // Displayes month and year (e.g. January | 2015) use case - Credit card expiry
        case DateAndMonth   // Displayes date and month (e.g. 28 | May). Use case - Anniversary
    }

    // default is UIDatePickerModeDateAndTime
    var datePickerMode: PSDatePickerMode

    private let monthArray: [String] = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
    
    private var day: Int = 0
    private var month: Int = 0
    private var year: Int = 0
    
    override init(frame: CGRect)
    {
        self.datePickerMode = .DateAndTime
        super.init(frame:frame)
    }
    
    required init(coder aDecoder: NSCoder) {

        self.datePickerMode = .DateAndTime
        super.init(coder: aDecoder)
    }
    
    convenience init(datePickerMode pickerMode: PSDatePickerMode)
    {
        switch pickerMode {
        case .MonthAndYear:
            
            var picker: UIPickerView = UIPickerView()
            picker.tag = 1002
            var frame: CGRect = picker.frame
            self.init(frame: frame)
            initialiseDateComponents(NSDate())
            day = 0
            self.datePickerMode = .MonthAndYear
            picker.dataSource = self
            picker.delegate = self
            picker.selectRow(currentMonthIndex(), inComponent: 0, animated: true)
            picker.selectRow(currentYearIndex(), inComponent: 1, animated: true)
            self.addSubview(picker)
            
        case .DateAndMonth:

            var picker: UIPickerView = UIPickerView()
            picker.tag = 1002
            var frame: CGRect = picker.frame
            self.init(frame: frame)
            initialiseDateComponents(NSDate())
            self.datePickerMode = .DateAndMonth
            year = 0
            picker.dataSource = self
            picker.delegate = self
            picker.selectRow(todayDateIndex(), inComponent: 0, animated: true)
            picker.selectRow(currentMonthIndex(), inComponent: 1, animated: true)
            self.addSubview(picker)
            
        default:
            
            var picker: UIDatePicker = UIDatePicker()
            var frame: CGRect = picker.frame
            self.init(frame: frame)
            picker.tag = 1002
            picker.datePickerMode = getUIDatePickerMode(pickerMode)
            self.addSubview(picker)
        }
    }
    
    convenience override init()
    {
        self.init(datePickerMode: .DateAndTime)
    }
    
    func description() -> String
    {
        var dateString = ""
        if self.datePickerMode == .DateAndMonth || self.datePickerMode == .MonthAndYear
        {
            if year != 0 {
                dateString += (dateString == "") ? "\(year)" : "-\(year)"
            }
            if month != 0 {
                var monthString: String = (month < 10) ? "0\(month)" : "\(month)"
                dateString += (dateString == "") ? monthString : "-\(monthString)"
            }
            if day != 0 {
                var dayString: String = (day < 10) ? "0\(day)" : "\(day)"
                dateString += (dateString == "") ? dayString : "-\(dayString)"
            }
            

        }else {
            var datePicker: UIDatePicker = self.viewWithTag(1002) as UIDatePicker
            var dateFormatter: NSDateFormatter = NSDateFormatter()
            dateString = datePicker.date.description
        }
        return dateString
    }
    
    func initialiseDateComponents(date: NSDate!){

        //getting date components for today's date
        let gregorianCalender: NSCalendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
        let flags: NSCalendarUnit = .CalendarUnitYear | .CalendarUnitMonth | .CalendarUnitDay
        let dateComponents: NSDateComponents = gregorianCalender.components(flags, fromDate: NSDate())
        
        day = dateComponents.day
        month = dateComponents.month
        year = dateComponents.year

    }
    //MARK: - UIPickerViewDataSource
    
    // returns the number of 'columns' to display.
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int
    {
        return 2
    }
    
    // returns the number of rows in each component..
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int
    {
        var number: Int = 0
        if self.datePickerMode == .DateAndMonth
        {
            if component == 0
            {
                number = [31,29,31,30,31,30,31,31,30,31,30,31][month]
            }else
            {
                number = 12     // 12 months
            }
        }else if self.datePickerMode == .MonthAndYear
        {
            if component == 0
            {
                number = 12     //12 months
            }else
            {
                number = 10000  //10000 year. Same as default UIDatePicker if maximum and minimum date is not set
            }
        }
        return number
    }

    //
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int){

        if self.datePickerMode == .DateAndMonth
        {
            if(component == 0)
            {
                //Date component is at 0
                day = row+1
                
            }else{
                //Month component is at 1
                month = row+1
                pickerView.reloadAllComponents()
            }
            
        }else if (self.datePickerMode == .MonthAndYear){
            if(component == 0)
            {
                //Month component is at 0
                month = row+1
            }else{
                //Year component is at 1
                year = row+1
            }

        }
    }
    
    //MARK: - UIPickerViewDelegate
    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String!
    {
        switch self.datePickerMode {
            
        case .DateAndMonth:
            return component == 0 ? "\(row+1)" : self.monthArray[row]
        case .MonthAndYear:
            return component == 0 ? self.monthArray[row] : "\(row+1)"
        default:
            return ""
        }
    }
    
    //MARK: - Helper methods
    func todayDateIndex() -> Int
    {
        let gregorianCalender: NSCalendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
        let flags: NSCalendarUnit = NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitMonth
        let dateComponents: NSDateComponents = gregorianCalender.components(flags, fromDate: NSDate())
        return dateComponents.day-1
    }
    
    func currentMonthIndex() -> Int
    {
        let gregorianCalender: NSCalendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
        let flags: NSCalendarUnit = NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitYear
        let dateComponents: NSDateComponents = gregorianCalender.components(flags, fromDate: NSDate())
        return dateComponents.month-1
    }
    
    func currentYearIndex() -> Int
    {
        let gregorianCalender: NSCalendar = NSCalendar(calendarIdentifier: NSGregorianCalendar)!
        let flags: NSCalendarUnit = NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitYear
        let dateComponents: NSDateComponents = gregorianCalender.components(flags, fromDate: NSDate())
        return dateComponents.year-1
    }
    
    func getUIDatePickerMode(datePickerMode: PSDatePickerMode) -> UIDatePickerMode {
        switch (datePickerMode) {
        case .Time:
            return UIDatePickerMode.Time
        case .Date:
            return UIDatePickerMode.Date
        case .DateAndTime:
            return UIDatePickerMode.DateAndTime
        case .CountDownTimer:
            return UIDatePickerMode.CountDownTimer
        default:
            return UIDatePickerMode.Date
        }
    }
    
    // if animated is YES, animate the wheels of time to display the new date
    func setDate(date: NSDate, animated: Bool){
        
    }

}